oda_api/api.py

```diff
@@ -1339,3 +1340,48 @@ def from_response_json(cls, res_json, instrument, product):
                 p.meta_data = p.meta
 
         return d
+
+class ProgressReporter(object):
+    """
+    The class allows to report task progress to end user
+    """
+    def __init__(self):
+        self._callback = None
+        callback_file = ".oda_api_callback"  # perhaps it would be better to define this constant in a common lib
+        if not os.path.isfile(callback_file):
+            return
+        with open(callback_file, 'r') as file:
+            self._callback = file.read().strip()
+
+    @property
+    def enabled(self):
+        return self._callback is not None
+
+    def report_progress(self, stage: str=None, progress: int=50, substage: str=None, subprogress: int=None, message:str=None):
+        """
+        Report progress via callback URL
+        :param stage: current stage description string
+        :param progress: current stage progress in %
+        :param substage: current substage description string
+        :param subprogress: current substage progress in %
+        :param message: message to pass
+        """
+        callback_payload = {k: str(v) for k, v in locals().items() if v is not None and k != 'self'}
+
+        if not self.enabled:
+            logger.info('no callback registered, skipping')
+            return
+
+        logger.info('will perform callback: %s', self._callback)
+
+        if re.match('^file://', self._callback):
+            with open(self._callback.replace('file://', ''), "w") as f:
+                json.dump(callback_payload, f)
+            logger.info('stored callback in a file %s', self._callback)
+
+        elif re.match('^https?://', self._callback):
+            r = requests.get(self._callback, params=callback_payload)
```

Comment:
Need try-except here to account for the situation when url is unreachable

Reply:
What do you mean @dsavchenko , do what with except?
If internal exception raises it can be shown as is, no need to re-raise.
Maybe a retry can be useful, but through requests, not custom.

Reply:
I mean this progress-reporting callback is not critical for the workflow itself, so why not log an error and allow workflow to continue?

Reply:
If communication with the dispatcher is not possible,  "done" message will not arrive and the user experience is "hanging request" which is bad.
Silently failing user-requested code throughout the execution is also in general unfortunate.
Actually I think this should end up in a sentry message, can we do it?

Reply:
This is not about "done" status, but progress report. "done" is sent by nb2workflow upon completion. The lack of intermediate progress report is not so critical. 

Sentry is an option, but then we need to pass sentry credentials to the backend, which runs user-provided code, this is a security issue

Reply:
> This is not about "done" status, but progress report. "done" is sent by nb2workflow upon completion. The lack of intermediate progress report is not so critical.

The most realistic reason it would fail is due to problem accessing dispatcher. The same problem would prevent "done".
If there is some bug, or the callback reference file is overwritten, or something like that - the functionality intended by the user will working, and the right course of action is to let us and the user know, not silently ignore. If there is a warning inside, some users will see some dropped progress messages and it will be hard to know when this happens. That's not a good experience.

> Sentry is an option, but then we need to pass sentry credentials to the backend, which runs user-provided code, this is a security issue

It is really not https://docs.sentry.io/product/sentry-basics/concepts/dsn-explainer/#dsn-utilization
The problem can be with alert flooding, which may be challenging in some cases, but it is not the case for us at the moment.
Note that every web tracking framework, e.g. google analytics, has a similar issue.

Reply:
We can probably keep it "resolved" here as the basic implementation, but I added an Issue https://github.com/oda-hub/oda_api/issues/229